Changes to episode 2 (CWL and Shell Tools) (#53)
carpentries-incubator/cwl-novice-tutorial: _episodes/02-shell_to_cwl.md

> and __define the files that will be included as output of a workflow__.
{: .callout}

> CWL workflows are defined in a YAML script. This [quick tutorial](https://www.commonwl.org/user_guide/yaml/) explains the YAML script used in CWL. A CWL script contains the workflow and the requirements for running that workflow. All CWL scripts should start with two lines of code:

Reviewer: ```suggestion
CWL workflows are written in the YAML syntax. This [short tutorial](https://www.commonwl.org/user_guide/yaml/)
explains the parts of YAML used in CWL. A CWL document contains the workflow and the
requirements for running that workflow. All CWL document should start with two lines of code:
```
- line breaks
- CWL document, not CWL script

~~~
cwlVersion: v1.2
class: 
~~~
{: .language-yaml}

> The `cwlVersion` string defines which standard of the language is required for the tool or workflow. The most recent version is v1.2, and defaulting to this will enable your scripts to use all versions of the language, though some workflow engines which are not up-to-date may not run the script. This is, however, a hurdle to be tackled when you reach it.

Reviewer: ```suggestion
The `cwlVersion` string defines which standard of the language is required for the tool or workflow. The most recent version is v1.2.
```
All the major implementations support v1.2, so lets not worry about the earlier versions of CWL.

> The `class` field defines what this particular script is. The majority of CWL scripts will fall into one of two classes: `CommandLineTool`, or `Workflow`. The `CommandLineTool` class is used for describing the interface for a command-line tool, while the `Workflow` class is used for collecting those tool descriptions into a workflow. In this lesson we will learn the differences between these two classes, how to pass data to and from command-line tools and specify working environments for these, and finally how to use a tool descriptor within a workflow.

Reviewer: ```suggestion
The `class` field defines what this particular document is. The majority of CWL documents will fall into
 one of two classes: `CommandLineTool`, or `Workflow`. The `CommandLineTool` class is used for
describing the interface for a command-line tool, while the `Workflow` class is used for connecting those
tool descriptions into a workflow. In this lesson we will learn the differences between these two classes,
how to pass data to and from command-line tools and specify working environments for these, and finally
how to use a tool description within a workflow.
```

## Our first CWL script

> To demonstrate the basic requirements for a tool descriptor we will recreate the standard "Hello world!" example.

Reviewer: ```suggestion
To demonstrate the basic requirements for a tool descriptor we will create a CWL description for the popular "Hello world!" demonstration.
```

__echo.cwl__
~~~

> clwVersion: v1.2

Reviewer: ```suggestion
cwlVersion: v1.2
```

class: CommandLineTool

baseCommand: echo

inputs:
  message_text:
    type: string
	inputBinding:
	  position: 1

outputs: []
~~~
{: .language-yaml}

Next, we create our input file: `hello_world.yml`.

__hello_world.yml__
~~~
message_text: Hello world!
~~~
{: .language-yaml}

> Now we can run our CWL script with the `cwltool`. This command needs our `cwl` workflow file along with our input file.

Reviewer: ```suggestion
Now we can run our CWL script with the reference CWL runner, `cwltool`. This command needs our `.cwl` workflow file along with our input file.
```

~~~
cwltool echo.cwl hello_world.yml
~~~
{: .language-bash}

~~~
INFO .../cwltool
INFO Resolved 'echo.cwl' to 'file:///.../echo.cwl'
INFO [job echo.cwl] /private/tmp/docker_tmprm65mucw$ echo \
    'Hello world!'
Hello world!
INFO [job echo.cwl] completed success
{}
INFO Final process status is success
~~~
{: .output}

The output displayed above shows us that the program has run succesfully and what its output is, `Hello world!`. 

Let's take a look at the `echo.cwl` script in more detail.

As explained above, the first 2 lines are always the same. You define the CWL version and which class you are using. In this example we have a `CommandLineTool`, in particular the `echo` command.
The next line, `baseCommand`, contains the command we want to run (`echo`). 

~~~
inputs:
  message_text:
    type: string
	inputBinding:
	  position: 1
~~~
{: .language-yaml}

This block of code contains the `inputs` section of our workflow. In this section we provide all the inputs that are needed for the workflow. 
In this example a string is needed that will be printed on the command line. We first define an id for our input: `message_text`. 

> Then we need to define the type of the input. In our case it will be a `string`. The `InputBinding` describes how the input should appear on the command line.

Reviewer: Did "we" write this? Since we didn't, maybe switch around the wording

Author: I've changed this in this commit: https://github.com/carpentries-incubator/cwl-novice-tutorial/pull/53/commits/86a46d2efaf811c46c7ef3b4e80fbed1783156cf

Reviewer 2: I must admit my writing style here was more informal/conversational - this does follow the general style that the software carpentry courses use though, e.g.: https://swcarpentry.github.io/shell-novice/02-filedir/index.html

I'm agnostic on which is a better style to use though. The software carpentry material is intended to be read out by an instructor to a class, so perhaps in that environment, conversational language is more fitting. I guess it depends on how you see the material being used, and who your target audience is?

Reviewer: Repeating what I discussed with @ALuesink elsewhere, this particular section is about reading an already written CWL tool description, not actively writing a new one. So no action verb phrases "we first define"; nothing is being written from scratch.

I am fully supportive of a conversational tone over a formal tone! :-)

> The `position` field indicates at which position the input will be on the command line, in our example at 1.

Reviewer: ```suggestion
The `position` field indicates at which position the input will be on the command line, in our example it is at the first position.
```

~~~
outputs: []
~~~
{: .language-yaml}

> Lastly the `outputs` of our workflow. In this example we don't have a formal output. Our text is printed directly on the command line. So we use an empty list (`[]`) as our output. 

Reviewer: ```suggestion
Lastly the `outputs` of our workflow. In this example we don't have a formal output.
Our text is printed directly in the terminal. So we use an empty YAML list (`[]`) as our output. 
```

> and __define the files that will be included as output of a workflow__.
{: .callout}

CWL workflows are written in the YAML syntax. This [short tutorial](https://www.commonwl.org/user_guide/yaml/)
explains the parts of YAML used in CWL. A CWL document contains the workflow and the

> requirements for running that workflow. All CWL document should start with two lines of code:

Reviewer: ```suggestion
requirements for running that workflow. All CWL documents should start with two lines of code:
```

~~~
cwltool echo.cwl hello_world.yml
~~~
{: .language-bash}

~~~
INFO .../cwltool
INFO Resolved 'echo.cwl' to 'file:///.../echo.cwl'
INFO [job echo.cwl] /private/tmp/docker_tmprm65mucw$ echo \
    'Hello world!'
Hello world!
INFO [job echo.cwl] completed success
{}
INFO Final process status is success
~~~
{: .output}

The output displayed above shows that the program has run succesfully and its output, `Hello world!`. 

Let's take a look at the `echo.cwl` script in more detail.

As explained above, the first 2 lines are always the same, the CWL version and the class of the script are defined. 
In this example the class is `CommandLineTool`, in particular the `echo` command.
The next line, `baseCommand`, contains the command that will be run (`echo`). 

~~~
inputs:
  message_text:
    type: string
	inputBinding:
	  position: 1
~~~
{: .language-yaml}

This block of code contains the `inputs` section of the workflow. This section provides all the inputs that are needed for the workflow. 
In this example a string is needed that will be printed on the command line. An id is defined first for the input: `message_text`. 

> Then the type of the input needs to be defined. In this case it will be a `string`. The `InputBinding` describes how the input should appear on the command line.

Reviewer: ```suggestion
Then the type of the input needs to be defined. In this case it will be a `string`. The `inputBinding` describes how the input should appear on the command line.
```

~~~
{: .language-yaml}

The `cwlVersion` string defines which standard of the language is required for the tool or workflow. The most recent version is v1.2.

The `class` field defines what this particular document is. The majority of CWL documents will fall into
one of two classes: `CommandLineTool`, or `Workflow`. The `CommandLineTool` class is used for
describing the interface for a command-line tool, while the `Workflow` class is used for connecting those
tool descriptions into a workflow. In this lesson the differences between these two classes are explained,
how to pass data to and from command-line tools and specify working environments for these, and finally
how to use a tool description within a workflow.

## Our first CWL script

To demonstrate the basic requirements for a tool descriptor a CWL description for the popular "Hello world!" demonstration will be examined.

__echo.cwl__
~~~
cwlVersion: v1.2
class: CommandLineTool

baseCommand: echo

inputs:
  message_text:
    type: string
	inputBinding:
	  position: 1

outputs: []
~~~
{: .language-yaml}

Next, the input file: `hello_world.yml`.

__hello_world.yml__
~~~
message_text: Hello world!
~~~
{: .language-yaml}

> The reference CWL runner, `cwltool`, is used to run the CWL script. This command needs the `.cwl` workflow file along with the `.yml` input file.

Reviewer: ```suggestion
We will use the reference CWL runner, `cwltool` to run this CWL document (the `.cwl` workflow file) along with the `.yml` input file.
```

~~~
cwltool echo.cwl hello_world.yml
~~~
{: .language-bash}

~~~

> INFO .../cwltool

Reviewer: ```suggestion
```

INFO [job echo.cwl] /private/tmp/docker_tmprm65mucw$ echo \
    'Hello world!'
Hello world!
INFO [job echo.cwl] completed success
{}
INFO Final process status is success
~~~
{: .output}

The output displayed above shows that the program has run succesfully and its output, `Hello world!`. 

Let's take a look at the `echo.cwl` script in more detail.

As explained above, the first 2 lines are always the same, the CWL version and the class of the script are defined. 
In this example the class is `CommandLineTool`, in particular the `echo` command.
The next line, `baseCommand`, contains the command that will be run (`echo`). 

~~~
inputs:
  message_text:
    type: string
	inputBinding:
	  position: 1
~~~
{: .language-yaml}

This block of code contains the `inputs` section of the workflow. This section provides all the inputs that are needed for the workflow. 
In this example a string is needed that will be printed on the command line. An id is defined first for the input: `message_text`. 
Then the type of the input needs to be defined. In this case it will be a `string`. The `inputBinding` describes how the input should appear on the command line.

> The `position` field indicates at which position the input will be on the command line, in this example it is at the first position.

Reviewer: ```suggestion
This block of code contains the `inputs` section of the tool description. This section provides all the inputs that are needed for running this specific tool. 
To run this example we will need to provide a string which will be included on the command line. Each of the inputs has a name, to help us tell them apart; this first input has the name : `message_text`. 
The field `inputBinding` is one way to specify how the input should appear on the command line.
Here the `position` field indicates at which position the input will be on the command line; in this case the `message_text` value will be the first thing added to the command line (after the `baseCommand`, `echo`).
```

~~~
outputs: []
~~~
{: .language-yaml}

> Lastly the `outputs` of the workflow. This example doesn't have a formal output.

Reviewer: ```suggestion
Lastly the `outputs` of the tool description. This example doesn't have a formal output.
```

The text is printed directly in the terminal. So an empty YAML list (`[]`) is used as the output. 

> ## Script order
> To make the script more readable the `input` field is put in front of the `output` field. 
> However CWL syntax requires only that each field is properly defined, it does not require them to be in a particular order. 
{: .callout}

> ## Changing input text
>

> > What do you need to change to print another text on the command line?

Reviewer: ```suggestion
> What do you need to change to print different text on the command line?
```

>
> > ## Solution
> >
> > To change the text on the command line, you only have to change the text in the `hello_world.yml` file.
> > 
> > For example:
> > ~~~
> > message_text: Good job!
> > ~~~
> > {: .language-yaml}
> {: .solution}
{: .challenge}

## CWL single step workflow

Next the RNA-seq data will be used for the first CWL workflow. The first step of RNA-sequensing analysis is a quality control of the RNA reads using `fastqc`.

> This tool is already available to use so there is no need to write a new CWL tool description.

Reviewer: ```suggestion
The RNA-seq data from the introduction episode will be used for the first CWL workflow. The first step of RNA-sequencing analysis is a quality control of the RNA reads using `fastqc`.
This tool is already available to use so there is no need to write a new CWL tool description.
```

This is the workflow file (`rna_seq_workflow.cwl`).

__rna_seq_workflow.cwl__
~~~
clwVersion: v1.2
class: Workflow

inputs:
  fastq: File
  
steps:
  fastqc:
    run: bio-cwl-tools/fastqc/fastqc_2.cwl
	in:
	  reads_file: fastq
    out: [html_file]

outputs: 
  qc_html:
    type: File
	outputSource: fastqc/html_file
~~~
{: .language-yaml}

In a __workflow__ the `steps` field must always be present. The workflow tasks or steps that you want to run are listed in this field. 
At the moment the workflow only contains one step: `fastqc`. In the next episodes more steps will be added to the workflow. 

> 

Reviewer: Lets give this step a function-based name, not a tool-based name

Author: Changed step name to `quality_control` in commit https://github.com/carpentries-incubator/cwl-novice-tutorial/pull/53/commits/56c8979c6c90b8bacf632dc2ca85e975191e8f85

Let's take a closer look at the workflow. First the `inputs` field will be explained.

~~~
inputs:
  fastq: File
~~~
{: .language-yaml}

Looking at the CWL script of the `fastqc` tool, it needs a fastq file as its input. So this workflow also needs a `File` for its inputs. 
To make this workflow interpretable for other researchers, self-explanatory and sensible variable names are used.

> In this case, `fastq` is used as the input name for the fastq file.

Reviewer: Again, lets choose a better name here. Is the final workflow oriented towards human sequencing data?

Author: Changed variable name to `fastq_file` in commit https://github.com/carpentries-incubator/cwl-novice-tutorial/pull/53/commits/56c8979c6c90b8bacf632dc2ca85e975191e8f85

## CWL single step workflow

Next the RNA-seq data will be used for the first CWL workflow. The first step of RNA-sequensing analysis is a quality control of the RNA reads using `fastqc`.

> This tool is already available to use so there is no need to write a tool descriptor.

Reviewer: ```suggestion
This tool is already available to use so there is no need to write a new CWL tool description.
```
